internal/fieldmask/fieldmask.go

```diff
@@ -0,0 +1,165 @@
+package fieldmask
+
+import (
+	"context"
+	"net/http"
+	"strings"
+
+	jsoniter "github.com/json-iterator/go"
+	"github.com/tidwall/gjson"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+)
+
+const key = "fields"
+
+// FieldMask is recursive structure to define a path mask
+type FieldMask map[string]FieldMask
+
+// New create a FieldMask from the input array of paths. The array should contain JSON paths with dit "." notation.
+func New(paths []string) FieldMask {
+	mask := make(FieldMask)
+	for _, path := range paths {
+		current := mask
+		fields := strings.Split(path, ".")
+		for _, field := range fields {
+			c, ok := current[field]
+			if !ok {
+				c = make(FieldMask)
+				current[field] = c
+			}
+			current = c
+		}
+	}
+	return mask
+}
+
+// Filter takes a Proto message as input and updates the message according to the FieldMask.
+func (fm FieldMask) Filter(message proto.Message) {
+	if len(fm) == 0 {
+		return
+	}
+
+	reflect := message.ProtoReflect()
+	reflect.Range(func(fd protoreflect.FieldDescriptor, v protoreflect.Value) bool {
+		mask, ok := fm[string(fd.Name())]
+		if !ok {
+			reflect.Clear(fd)
+		}
+
+		if len(mask) == 0 {
+			return true
+		}
+
+		switch {
+		case fd.IsMap():
+			m := reflect.Get(fd).Map()
+			m.Range(func(k protoreflect.MapKey, v protoreflect.Value) bool {
+				if fm, ok := mask[k.String()]; ok {
+					if i, ok := v.Interface().(protoreflect.Message); ok && len(fm) > 0 {
+						fm.Filter(i.Interface())
+					}
+				} else {
+					m.Clear(k)
+				}
+				return true
+			})
+		case fd.IsList():
+			list := reflect.Get(fd).List()
+			for i := 0; i < list.Len(); i++ {
+				mask.Filter(list.Get(i).Message().Interface())
+			}
+		case fd.Kind() == protoreflect.MessageKind:
+			mask.Filter(reflect.Get(fd).Message().Interface())
+		case fd.Kind() == protoreflect.BytesKind:
+			if b := v.Bytes(); gjson.ValidBytes(b) {
+				b, err := jsoniter.Marshal(mask.FilterJSON(b, []string{}))
+				if err == nil {
+					reflect.Set(fd, protoreflect.ValueOfBytes(b))
+				}
+			}
+		}
```

Comment:
feels like we should have a default case to at least log unexpected messages, no?

Reply:
I would say no, because this is simply filtering. Any unknown cases will just be ignored.
And for logging, there's are some restrictions here. We cannot throw `error` because then the client will receive the  error. We cannot bring logger from the context, then the implementation will have dependency and can't be used as a library.
Although we can still log with normal golang's `fmt` or `logger` library, but that will break structured logging.

Reply:
valid reasoning 

I still wonder about being able to debug in the future "why did my filter not catch 'X'"

could we emit events for the errors in this switch?

could we emit events for the default case?  Or are there too many types besides MessageKind and BytesKind?

Reply:
yes k8s events .... api server has a k8s client so in theory could emit events to help with debug, and you don't have the logging pain you noted earlier